### ISSUE TITLE
Negative HEAT_TRANS_COEF

### ISSUE THREAD

**Issue by rmcdermo** [firemodels/fds#10593]
Try this case, and look at the PARTICLE HEAT TRANSFER COEFFICIENT.  As you say, this may only happen for cases with small DTMP, but it makes the SMV output annoying.  To trap the negative values I used this:
```
IF (ABS(DTMP)>TWO_EPSILON_EB .AND. ABS(ONE_D%Q_CON_F)>TWO_EPSILON_EB) THEN
   ONE_D%HEAT_TRANS_COEF = ONE_D%Q_CON_F/DTMP
ELSE
   ONE_D%HEAT_TRANS_COEF = HTCF
ENDIF

if (ONE_D%HEAT_TRANS_COEF<0._EB) write(lu_err,*) 'ht1d', DTMP, ONE_D%Q_CON_F, ONE_D%HEAT_TRANS_COEF
```
near the end of SOLID_HEAT_TRANSFER_1D.

Here is the input file.

[tree_2_m_14_pc.fds.txt](https://github.com/firemodels/fds/files/8689388/tree_2_m_14_pc.fds.txt)

<img width="752" alt="Screen Shot 2022-05-13 at 12 23 07 PM" src="https://user-images.githubusercontent.com/4418497/168326216-38f45b47-75ff-47d9-987b-aae9d88aa154.png">

**Comment by mcgratta** [comment]
I suggest that `ONE%D_HEAT_TRANS_COEF` take whatever its last calculated value was from the 1-D heat conduction routine. It is used to calculate things like adiabatic surface temperature, so I think it is better that it not be a negative number.

**Comment by drjfloyd** [comment]
In the 1D HT routine we do a check to see if we need smaller timesteps.  In each subtimestep HTCF is computed using TMP_F at the start of that subtimestep.  Over the subtimesteps HTCF changes. This is why we need to recompute HTCF since the final value of HTCF after multiple subtimesteps is not going to match the effective HTCF over all the subtimesteps.   We added this logic to avoid pulling our hair out when debugging issues with the wall heat transfer. If we don't do this then the output h written to devc can't be used to check the convective flux.

In this case particles need smaller timesteps. What I see happening (looking at particle 83) is the particle starts of slightly cooler than the gas. Then with each subtimestep the particle heats until at the end the particle has a higher temperature than the gas.  With the Crank Nicholson time step we have qconf = h (TMPG - 0.5 (TMP_F_0 + TMP_F_1)). In this case the average of TMP_F_1 is slightly lower than the gas temp but the net heat transfer is from the particle to the gas. This results in the negative h.  

We could if h goes negative set it to the last HTCF. This is an odd situation that I assume results from the reaction causing a temperature rise above the gas but not high enough to avoid the negative h. I wouldn't want that to be the default behavior when h is positive as that would impact debugging using devc.

**Comment by drjfloyd** [comment]
or we just take the max of 0 and the HTC when doing AST

**Comment by rmcdermo** [comment]
Jason,
Which case gives you are hard time with DEVC using HTCF?

**Comment by drjfloyd** [comment]
There have been multiple occasions in the past (not necessarily issues on the tracker) where trying to figure out problems with wall heat transfer has meant running with DT_DEVC small enough to output the timestep data and using that to try and figure out what was going wrong.  Having the values in the devc file not match what is actually going on in FDS seems to always wind up causing us problems at some point.

**Comment by rmcdermo** [comment]
OK, I get that.  But I'm not ok with the output we have now.  I'd like to have a case that I can play with as I try so solve this problem.  It is hard to follow what Q_CON_F is doing in the Crank-Nicolson routine.  Sometimes it is computed using qconf=h(TG-TF), sometimes it's computed with qconf=h(TG-.5(TFOLD-TF)), etc.  Other times it seems to just be a value placeholder.

Please try to design an example case that requires the effective h and qconf to be matched up beyond just using HTCF.  I'd like to see if we can have our cake and it too by having a qconf that satisfies Crank-Nicolson while also an h with reasonable values.  Thanks

**Comment by drjfloyd** [comment]
Q_CON_F is just h (TG-0.5*(T_F_OLD+T_F).

In the subtimestep routine this is broken into two parts:
We start with TMP_F being the beginning of timestep value.
The first part is using this:
Q_CON_F  = (TG-0.5*TMP_F)*DT_BC_SUB 
then we update TMP_F with the new ONE_D%TMP values
The second part is the second half of the Crank-Nicholson:
Q_CON_F = Q_CON_F - 0.5*TMP_F*DT_BC_SUB

then after finishing any subtimestepping to get back to a flux we do Q_CON_F = Q_CON_F / DT_BC

In this case here if you were to write QCONF and h for the particles to a devc file you would wind up with a timestep where the particle took a dozen substeps and the numbers were like (just from memory I'd have to put the debug back in and rerun):
TMP_G = 293.4
TMP_F_OLD = 293.2
TMP_F = 293.5
Q_CONF = -2
H = 30

If you were to go check Q_CONF with H and the TMP_F for the current and prior timestep you get 30 * (293.4 - 0.5(293.2+293.5)) = 1.5 and not -2.

**Comment by drjfloyd** [comment]
Changed to use the last value of HTCF computed.

**Comment by rmcdermo** [comment]
Latest change works.  I'll add EFFECTIVE HTC to output.

**Comment by drjfloyd** [comment]
Thanks.
